% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/shrinkage.R
\name{mr.raps.shrinkage}
\alias{mr.raps.shrinkage}
\alias{print.mr.raps}
\alias{plot.mr.raps}
\title{Main function for RAPS (shrinkage weights)}
\usage{
mr.raps.shrinkage(
  b_exp,
  b_out,
  se_exp,
  se_out,
  over.dispersion = FALSE,
  loss.function = c("l2", "huber", "tukey"),
  k = switch(loss.function[1], l2 = 2, huber = 1.345, tukey = 4.685),
  shrinkage = FALSE,
  prior.param = NULL,
  diagnostics = FALSE,
  se.method = c("sandwich", "bootstrap"),
  empirical.variability = FALSE,
  num.init = 10,
  multiple.root.warning = 1
)

\method{print}{mr.raps}(x, ...)

\method{plot}{mr.raps}(x, ...)
}
\arguments{
\item{b_exp}{A vector of SNP effects on the exposure variable, usually obtained from a GWAS.}

\item{b_out}{A vector of SNP effects on the outcome variable, usually obtained from a GWAS.}

\item{se_exp}{A vector of standard errors of \code{b_exp}.}

\item{se_out}{A vector of standard errors of \code{b_out}.}

\item{over.dispersion}{Should the model consider overdispersion (systematic pleiotropy)? Default is FALSE.}

\item{loss.function}{Either the squared error loss (\code{l2}) or robust loss functions/scores (\code{huber} or \code{tukey}).}

\item{k}{Threshold parameter in the Huber and Tukey loss functions.}

\item{shrinkage}{If shrinkage (empirical partially Bayes) should be used. Shrinkage does not affect the unbiasedness of the estimating equations and generally will increase the estimation accuracy. If TRUE, \code{prior.param} must be provided.}

\item{prior.param}{Parameters of the Gaussian spike-and-slab prior}

\item{diagnostics}{Should the function returns diagnostic plots and results? Default is FALSE}

\item{se.method}{How should the standard error be estimated? Either by sandwich variance formula (default and recommended) or the bootstrap.}

\item{empirical.variability}{When using the sandwich standard error, Whether the variability should be estimated by the empirical variance of the score.}

\item{num.init}{Number of initializations.}

\item{multiple.root.warning}{How to handle multiple roots of the estimating equations? When this happens, the results of \code{mr.raps.shrinkage} are less reliable. This parameter can take three values: 0---nothing will be done; 1---a warning is given; 2---an error is given. Default is 1.}

\item{x}{a \code{mr.raps} object}

\item{...}{further arguments (not supported)}
}
\description{
Main function for RAPS (shrinkage weights)
}
\details{
\code{mr.raps.shrinkage} is the main function for RAPS in conjunction with empirical partially Bayes. It is more general than the first generation \code{mr.raps.mle} function and should be preferred in practice. With the option \code{shrinkage = TRUE}, it essentially reduces to \code{mr.raps.mle}. In that case, the main difference is that the standard errors in \code{mr.raps.shrinkage} are computed based on observed information (and also an empirical estimate of the variance of the score function). This is preferred over using the plugged-in Fisher information in \code{mr.raps.mle}. See Efron and Hinkley (1978) referenced below.

Because the estimating equations are highly non-linear, it is possible that there are multiple roots. To overcome this issue, we use multiple initializations (controlled by \code{num.init}) around the \code{mr.raps.mle} point estimate. A warning is given if there seems to be another finite root, and no solution is returned if there are two roots close to the initialization. When the program does not find a finite solution, consider increasing the value of \code{num.init}.
}
\section{Methods (by generic)}{
\itemize{
\item \code{print}: Print

\item \code{plot}: Diagnostic plots
}}

\examples{

require(mr.raps)
data(lipid.cad)
data <- subset(lipid.cad, lipid == "hdl" & restrict &
gwas.selection == "teslovich_2010" &
gwas.outcome == "cardiogramplusc4d_1000genome")
z <- data$beta.exposure / data$se.exposure
prior.param <- fit.mixture.model(z)

## Results
mr.raps.shrinkage(data$beta.exposure, data$beta.outcome,
data$se.exposure, data$se.outcome, TRUE, "huber", shrinkage = FALSE)
\donttest{
mr.raps.shrinkage(data$beta.exposure, data$beta.outcome,
data$se.exposure, data$se.outcome, TRUE, "huber", shrinkage = TRUE,
prior.param = prior.param)
}

}
\references{
Qingyuan Zhao, Q., Chen, Y., Wang, J., and Small, D. S. (2018). A genome-wide design and an empirical partially Bayes approach to increase the power of Mendelian randomization, with application to the effect of blood lipids on cardiovascular disease. <arXiv:1804.07371>.
Efron, B. and Hinkley, D. V. (1978). Assessing the accuracy of the maximum likelihood estimator: Observed versus expected Fisher information. Biometrika, 65(3), 457--483.
}
